fix: timeout CAR fetch if no new block

willscott opened the pull request:
Timeout reads if no new block of a car has been received in X amount of time

Closes #91

lidel commented on lib/graph_gateway.go:
```diff
@@ -315,14 +316,46 @@ func (api *GraphGateway) loadRequestIntoSharedBlockstoreAndBlocksGateway(ctx con
+				}
+			}()
+
+			t := time.NewTimer(30 * time.Second)
```
:thought_balloon:    timeout for  `?format=raw` fetch across all retries  is 1m (`GetBlockTimeout`).
Switching to `?format=car` increased fetch timeout to 30m, so we are waiting longer than 1m for the first block.

Changing it to 30s means we will be failing to fetch the first block more often than the old GRAPH_BACKEND=false.
We could start with the same timeout as for `?format=raw` for now,

```suggestion
			t := time.NewTimer(GetBlockTimeout)
```

@willscott thoughts?

willscott replied:
makes sense 👍 changed here and below on line 354

aschmahmann reviewed (commented):
Generally seems good, a little nervous about our lack of tests in this area of the code in general. Have you done some manual testing to see how it goes?

BigLep commented:
> Generally seems good, a little nervous about our lack of tests in this area of the code in general. Have you done some manual testing to see how it goes?

What is causing the lack of testing?  Too hard to test?  Feeling rushed?  I want to make sure we do this right and not rely on manual testing alone.

willscott commented on lib/graph_gateway.go:
```diff
@@ -38,6 +39,8 @@ import (
 
 var graphLog = golog.Logger("backend/graph")
 
+const GetBlockTimeout = time.Second * 60
```
This is parallel to https://github.com/ipfs/bifrost-gateway/blob/16a33a8b695c7cb26a7651177227750e879d3335/blockstore.go#L21 - it's being duplicated here so as not to cause a circular dependency

lidel replied:
either way, having the same value twice under the same name feels error prone, and makes grepping harder.
we have only two files in  `lib`, and there are other "lib" things in the top directory -- unsure if the price of having `lib` is justified, maybe just remove this submodule and move things to top dir?

willscott commented on lib/graph_gateway.go:
```diff
@@ -315,14 +318,46 @@ func (api *GraphGateway) loadRequestIntoSharedBlockstoreAndBlocksGateway(ctx con
+					}
+					t.Reset(GetBlockTimeout)
+				case <-t.C:
+					return io.ErrNoProgress
```
Of the things I saw on `io` it seemed closest to why we were failing - is there another one you would suggest?

lidel replied:
Ideally use wrapped `gateway.ErrGatewayTimeout`, this way boxo/gateway library will translate such timeout to proper HTTP 504 (if its not handled anywhere sooner).

willscott commented:
@BigLep tests are probably doable with enough of a mocked server built up.

However this code is likely to change pretty soon with stronger validation logic, so I'm going to hold off investing more time here to see if it remains relevant after @aschmahmann 's next iteration.

willscott commented on lib/graph_gateway.go:
```diff
@@ -315,14 +318,46 @@ func (api *GraphGateway) loadRequestIntoSharedBlockstoreAndBlocksGateway(ctx con
+					if rdErr != nil {
+						err = rdErr
+					}
+					blkCh <- blk
```
the callback to `fetcher.Fetch` may be called multiple times, so i don't think we want to cancel the context in the callback

aschmahmann replied:
Thanks, that makes sense not to cancel in the callback then.

Do we need to add a cancellable context in the callback though along with a select statement around this line so that it won't just hang if a block shows up after `io.ErrNoProgress` is returned and we stop reading off the channel?

willscott commented on lib/graph_gateway.go:
```diff
@@ -315,14 +318,46 @@ func (api *GraphGateway) loadRequestIntoSharedBlockstoreAndBlocksGateway(ctx con
+				}
+			}()
+
+			t := time.NewTimer(GetBlockTimeout)
```
is your worry that we'd end up failing with `ErrNoProgress` when we should fail with the error surfaced from the timeout of `cr.Next()`?

we can make the initial timer 2x block timeout so that we get the underlying error in that case

aschmahmann replied:
My concern is that if we get sent back the CAR header but not the first block that we'll wait 30 minutes instead of 1. Am I misreading how this works?

aschmahmann replied:
Now it's just the header, which is fine that can have it's own timeout handled by the Fetch implementation. IIRC before the first block wasn't controlled by a block request timeout either, but it is now (with a 2x GetBlockTimeout).

willscott commented:
I believe I've addressed your comments, @aschmahmann  - i think this pattern is going to be reasonable for this class of timeouts, and something useful in addition to stricter validation. Let me know if there's a reason we want to not merge it?

aschmahmann commented:
@willscott I think this mostly looks good. If you could resolve https://github.com/ipfs/bifrost-gateway/pull/111#discussion_r1202430525 via a fix or explaining why it's ok as is then no issues from me.
